Libraries/LibWeb/CredentialManagement/CredentialsContainer.cpp

```diff
@@ -17,25 +20,476 @@ GC::Ref<CredentialsContainer> CredentialsContainer::create(JS::Realm& realm)
+        //      3. options is matchable a priori.
+        //      4. options.mediation is not "required".
+        //      5. options.mediation is not "conditional".
+        if (credentials.size() == 1
+            && is_matchable_a_priori(options)
+            && options.mediation != Bindings::CredentialMediationRequirement::Required
+            && options.mediation != Bindings::CredentialMediationRequirement::Conditional) {
+            WebIDL::resolve_promise(realm(), *promise, credentials[0]);
+            return;
+        }
+
+        // 4. If options’ mediation is "silent", resolve p with null, and skip the remaining steps.
+        if (options.mediation == Bindings::CredentialMediationRequirement::Silent) {
+            WebIDL::resolve_promise(realm(), *promise, JS::js_null());
+            return;
+        }
+
+        // 5. Let result be the result of asking the user to choose a Credential, given options and credentials.
+        auto result = ask_the_user_to_choose_a_credential(options, credentials);
+
+        // 6. If result is an interface object:
+        if (result.has<CredentialInterface*>()) {
+            // 1. Set result to the result of executing result’s [[DiscoverFromExternalSource]](origin, options, sameOriginWithAncestors),
+            //    given origin, options, and sameOriginWithAncestors.
+            auto maybe_result = result.get<CredentialInterface*>()->discover_from_external_source(realm(), origin, options, same_origin_with_ancestors);
+            // If that threw an exception:
+            if (maybe_result.is_error()) {
+                // 1. Let e be the thrown exception.
+                auto e = maybe_result.error_value();
+                // 2. Queue a task on global’s DOM manipulation task source to run the following substeps:
+                queue_global_task(HTML::Task::Source::DOMManipulation, settings.global_object(), GC::create_function(realm().heap(), [&] {
+                    HTML::TemporaryExecutionContext execution_context { realm(), HTML::TemporaryExecutionContext::CallbacksEnabled::Yes };
+
+                    // 1. Reject p with e.
+                    WebIDL::reject_promise(realm(), *promise, e);
+                }));
+                // 3. Terminate these substeps.
+                return;
+            }
+        }
+
+        // 7. Assert: result is null, or a Credential.
+        VERIFY(result.has<Empty>() || result.has<GC::Ref<Credential>>());
+
+        // 8. If result is a Credential, resolve p with result.
+        if (result.has<GC::Ref<Credential>>()) {
+            WebIDL::resolve_promise(realm(), *promise, result.get<GC::Ref<Credential>>());
+            return;
+        }
+
+        // 9. If result is null and options.mediation is not conditional, resolve p with result.
+        if (result.has<Empty>() && options.mediation != Bindings::CredentialMediationRequirement::Conditional)
+            WebIDL::resolve_promise(realm(), *promise, JS::js_null());
```

Comment:
am I missing something or is there a way to fall of the end of this parallel processing without resolving or rejecting the promise at all? If there is, that's probably a spec bug. If there isn't, we should VERIFY_NOT_REACHED() at the end.

Reply:
Yes, it appers that if `mediation` is conditional we'll fall through. Another issue :)

https://github.com/w3c/webappsec-credential-management/issues/271